### PR TITLE
Use iml-common from pypi and copr

### PR DESCRIPTION
**Pull request by tanabarr**
Currently *chroma-common* is not a python package but a subdirectory of this repository and packages link to it through symbolic links (chroma-agent and chroma-manager). *chroma-common*  has been extracted into a stand-alone python package https://github.com/intel-hpdd/iml-common/ and deployed to Pypi https://pypi.python.org/pypi/iml-common/ . This provides separation of the common code including relevant unit tests and provides capability to install, import and use the package directly rather than through symbolic links.

*chroma-common* can be removed from this repo and *iml-common* installed as a dependency from either copr (rpm) or pypi (with pip) instead in build , test and deployment.

### CODE REVIEW

File: chroma-agent/chroma-agent.spec [hunk 1]
```diff
@@ -28,6 +28,7 @@ Requires: python2-tablib
 Requires: yum-utils
 Requires: initscripts
 Requires: chroma-diagnostics >= %{version}
+Required: python-iml-common
```
Comment (brianjmurrell) [inline chroma-agent/chroma-agent.spec]: Typo: s/d:/s:/

File: chroma-manager/chroma-manager.spec [hunk 2]
```diff
@@ -113,6 +112,7 @@ This is the Intel Manager for Lustre Monitoring and Administration Interface
 %package libs
 Summary: Common libraries for Chroma Server
 Group: System/Libraries
+Requires: python-iml-common
```
Comment (brianjmurrell) [inline chroma-manager/chroma-manager.spec]: Should use python2-iml-common here.

File: chroma-manager/chroma-manager.spec [hunk 3]
```diff
@@ -127,7 +127,7 @@ or on a separate node.
 %package integration-tests
 Summary: Intel Manager for Lustre Integration Tests
 Group: Development/Tools
-Requires: python-requests >= 2.6.0 python-nose python-nose-testconfig python-paramiko python-django python-ordereddict
+Requires: python-requests >= 2.6.0 python-nose python-nose-testconfig python-paramiko python-django python-ordereddict python-iml-common
```
Comment (brianjmurrell) [inline chroma-manager/chroma-manager.spec]: Ditto python2-iml-common

File: chroma-manager/Makefile [hunk 4]
```diff
@@ -208,6 +208,13 @@ rpms: cleandist tarball
 	rm -rf _topdir
 	echo "jenkins_fold:end:Make Manager RPMS"
 
+install_requirements: requirements
+	echo "jenkins_fold:start:Install Python requirements"
+	pip install --upgrade pip;                              \
+	pip install --upgrade setuptools;                       \
+	pip install -r requirements.txt
+	echo "jenkins_fold:end:Install Python requirements"
+
```
Comment (brianjmurrell) [inline chroma-manager/Makefile]: [Here's a technique](http://blog.bottlepy.org/2012/07/16/virtualenv-and-makefiles.html) to avoid doing all of that virtualenv configuration (i.e. every single time even) if the `requirements.txt` file has not actually changed.

Comment (tanabarr) [inline chroma-manager/Makefile]: if you've already got a patch , then I will leave as is and we can fix subsequently, no point implementing the same in 2 different patches, this is just a minor DRY improvement

Comment (brianjmurrell) [inline chroma-manager/Makefile]: My patch applies to the top-level `Makefile`.  You want the similar technique in the `chroma-manager/Makefile`, so even when mine lands, it will not apply the technique for `chroma-manager/Makefile`.

So you still need that in this patch, but I'm just suggesting to use this technique to avoid wasting more time on each build/test run as the time to do builds and tests has been expressed as taking too long many times already.  That is partly because we waste time running through processes which end up being a NOOP which we can avoid.  This would be another one of those.

You can see the technique as I have applied it in the top-level `Makefile` [here](https://github.com/intel-hpdd/intel-manager-for-lustre/pull/34/commits/626489658c35e5125007e6bda3d56e3ea0692e74#diff-b67911656ef5d18c4ae36cb6741b7965R125)

File: chroma-manager/requirements.dev [hunk 5]
```diff
@@ -18,4 +18,4 @@ pyflakes
 pygraphviz
 sphinx==1.1.3
 werkzeug
-iml-common
+iml-common>=1.0.5
```
Comment (brianjmurrell) [inline chroma-manager/requirements.dev]: Just curious why we need to pin this?

Just because, I think generally speaking, we'd like to get away from pinning since we are so bad at updating the pins to something recent and we then get into a place where we are unknowingly using old, vulnerable versions of modules.

Comment (tanabarr) [inline chroma-manager/requirements.dev]: we need to pin here because the unit tests don't clear packages inside the virtual environment used, if you want to resolve that it would be much appreciated (as per previous discussion)

Comment (brianjmurrell) [inline chroma-manager/requirements.dev]: > we need to pin here because the unit tests don't clear packages inside the virtual environment used

But pinning isn't clearing is it?  I'm still not sure I understand what the problem is.

Doesn't `pip` always ensure the latest version is installed, upgrading an existing venv to the latest if it's not already?  Don't we always just want to use the latest version?

> if you want to resolve that it would be much appreciated

What problem(s) are resolved by clearing the venv and recreating it each time?  Maybe we need to get a ticket opened on this issue so that we can understand what the problem is and come up with a solution to whatever the problem is.

File: chroma-manager/Makefile [hunk 6]
```diff
@@ -208,6 +208,13 @@ rpms: cleandist tarball
...
 	echo "jenkins_fold:end:Make Manager RPMS"
 
+install_requirements: requirements
+	echo "jenkins_fold:start:Install Python requirements"
+	pip install --upgrade pip;                              \
+	pip install --upgrade setuptools;                       \
+	pip install -Ur requirements.txt
+	echo "jenkins_fold:end:Install Python requirements"
+
 requirements: requirements.txt
```
Comment (brianjmurrell) [inline chroma-manager/Makefile]: Why are we re-directing `requirements.txt` through `requirements` here?

It would be better to just remove this indirection and make all of the `requirements` dependencies `requirements.txt`

File: chroma-manager/Makefile [hunk 7]
```diff
@@ -208,6 +208,13 @@ rpms: cleandist tarball
 	rm -rf _topdir
 	echo "jenkins_fold:end:Make Manager RPMS"
 
+install_requirements: requirements
```
Comment (brianjmurrell) [inline chroma-manager/Makefile]: Should depend on `requirements.txt`.

File: chroma-manager/Makefile [hunk 8]
```diff
@@ -228,18 +235,17 @@ requirements.txt: ../chroma-dependencies/Makefile ../chroma-dependencies/*/Makef
 	fi
 	echo "jenkins_fold:end:Make requirements.txt"
 
-download: requirements.txt
-	echo "jenkins_fold:start:Install Python requirements"
-	pip install -r requirements.txt
-	echo "jenkins_fold:end:Install Python requirements"
+download: install_requirements
```
Comment (brianjmurrell) [inline chroma-manager/Makefile]: Should depend on `requirements.txt`.

File: chroma-manager/Makefile (hunk 7 above)
Comment (tanabarr) [inline chroma-manager/Makefile]: requirements depends on requirements.txt

File: chroma-manager/Makefile (hunk 6 above)
Comment (tanabarr) [inline chroma-manager/Makefile]: don't think it's necessary to be  repush for this , but I will change if repush for something

File: chroma-manager/Makefile (hunk 8 above)
Comment (tanabarr) [inline chroma-manager/Makefile]: requirements depends on requirements.txt

File: chroma-manager/Makefile (hunk 7 above)
Comment (brianjmurrell) [inline chroma-manager/Makefile]: Exactly.  A make rule such as `foo: bar` with no recipe is a NOOP with no function other than to just confuse.

File: chroma-manager/Makefile (hunk 8 above)
Comment (brianjmurrell) [inline chroma-manager/Makefile]: @tanabarr Yes, you are correct.  Sorry, I confused the concern about this target with the others.

But what is this `download` target doing?  It has no recipe, so why does it need to have the requirements installed?  Why is it even needed?

File: chroma-agent/tests/actions_plugins/test_manage_copytool.py [hunk 9]
```diff
@@ -53,67 +54,79 @@ def _configure_copytool(self):
                                        self.ct_arguments)
 
     def test_start_monitored_copytool(self):
-        self.single_commands(CommandCaptureCommand(('/sbin/service', 'chroma-copytool-monitor-%s' % self.ct_id, 'status'), rc=1),
-                             CommandCaptureCommand(('/sbin/service', 'chroma-copytool-monitor-%s' % self.ct_id, 'start')),
-                             CommandCaptureCommand(('/sbin/service', 'chroma-copytool-monitor-%s' % self.ct_id, 'status')),
-                             CommandCaptureCommand(('/sbin/service', 'chroma-copytool-%s' % self.ct_id, 'status'), rc=1),
-                             CommandCaptureCommand(('/sbin/service', 'chroma-copytool-%s' % self.ct_id, 'start')),
-                             CommandCaptureCommand(('/sbin/service', 'chroma-copytool-%s' % self.ct_id, 'status')))
+        self.single_commands(CommandCaptureCommand(('systemctl', 'daemon-reload')),
```
Comment (jgrund) [inline chroma-agent/tests/actions_plugins/test_manage_copytool.py]: How come these changed to systemctl ?

File: chroma-manager/tests/feature/cli/features/steps/standard_cli.py [hunk 10]
```diff
@@ -91,5 +91,5 @@ def step(context, message):
         ok_(message not in "".join(context.stdout.readlines()))
     except AssertionError:
         context.stdout.seek(0)
-        print context.stdout.readlines()
+        print(context.stdout.readlines())
```
Comment (jgrund) [inline chroma-manager/tests/feature/cli/features/steps/standard_cli.py]: Thought this was landed in another patch already?

File: chroma-manager/tests/integration/shared_storage_configuration/test_long_polling.py [hunk 11]
```diff
@@ -190,7 +191,7 @@ def test_long_polling_locks(self):
         self.long_polling_end_point.join()
 
         # Ensure update command completes before we exit the test
-        self.wait_for_command(self.chroma_manager, response.json['id'])
+        self.wait_for_command(self.chroma_manager, response.json['id'], timeout=LONG_TEST_TIMEOUT)
```
Comment (jgrund) [inline chroma-manager/tests/integration/shared_storage_configuration/test_long_polling.py]: Thought this had landed already?

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 12]
```diff
@@ -10,8 +10,67 @@
 import json
 import re
 import sys
-
-from chroma_common.lib.shell import Shell
```
Comment (jgrund) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: I feel like you've explained this before, but why does this need to be inlined?

File: chroma-manager/Makefile [hunk 13]
```diff
@@ -208,7 +208,12 @@ rpms: cleandist tarball
 	rm -rf _topdir
 	echo "jenkins_fold:end:Make Manager RPMS"
 
-requirements: requirements.txt
+install_requirements: requirements.txt
```
Comment (jgrund) [inline chroma-manager/Makefile]: Does this belong in the makefile if it updates the callers pip and setuptools?

File: chroma-agent/tests/actions_plugins/test_manage_copytool.py (hunk 9 above)
Comment (tanabarr) [inline chroma-agent/tests/actions_plugins/test_manage_copytool.py]: updated platform info mocking so now we default to el7 in unit tests

File: chroma-manager/tests/integration/utils/chroma_log_collector.py (hunk 12 above)
Comment (tanabarr) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: it gets used before we install chroma packages and before we install pip/virtual environments, this used to work because we had a symlink to the source of chroma-common, now that is no longer valid

File: chroma-manager/Makefile (hunk 13 above)
Comment (tanabarr) [inline chroma-manager/Makefile]: that is a valid concern, although the disadvantages of updating those critical packages are probably minimal
